LocalLang/Semantics.lean

```diff
@@ -1,12 +1,41 @@
 import LocalLang.AST
 
+inductive Ctx : Type where
+  | hole : Ctx
+  | bin_op_lhs : Ctx → BinOp → Expr → Ctx
+  | bin_op_rhs : ℕ → BinOp → Ctx → Ctx
+  | fun_call : String → List ℕ → Ctx → List Expr → Ctx
+
+def Ctx.fill (e : Expr) : Ctx → Expr
+  | hole => e
+  | bin_op_lhs ctx op e' => Expr.binOp op (ctx.fill e) e'
+  | bin_op_rhs n op ctx => Expr.binOp op (.const n) (ctx.fill e)
+  | fun_call f es_left ctx es_right =>
+      Expr.funCall f ((.const ·) <$> es_left ++ [ctx.fill e] ++ es_right)
+
+abbrev VarContext := Std.HashMap String ℕ
```

Comment:
To avoid overloading "context" yet again, perhaps this can better be called environment?